File: server/functions/src/accessors/users/syncs/user_sync_google_mail_accessor.ts
import {SemurEngineConfig} from "../../../config";
import {SemurEngineErrorBuilder} from "../../../semur_engine/semur_engine";
import {UserAccessor} from "../user_accessor";
import {SyncInformation} from "../../../models.pb/syncs/sync";
import {SyncGoogleMailEmail} from "../../../models.pb/syncs/sync_google-mail";
import {firestore} from "firebase-admin";
import Timestamp = firestore.Timestamp;

export enum UserSyncGoogleMailQueryKey {
    All = "all",
    CommonFilters = "commonFilters",
}

export interface UserSyncGoogleMailQueryArgs {
    [UserSyncGoogleMailQueryKey.All]: {
        userId: string;
        nangoIntegrationId: string;
        limit?: number;
    };
    [UserSyncGoogleMailQueryKey.CommonFilters]: {
        userId: string;
        nangoIntegrationId: string;
        from?: Timestamp;
        to?: Timestamp;
        threadId?: boolean;
        limit?: number;
    };
}

export class UserSyncGoogleMailAccessor {
    static firestoreCollectionName = "syncs";
    static firestoreSyncDocumentId = "sync";
    static firestoreUsersCollection = SemurEngineConfig.db.collection(UserAccessor.firestoreCollectionName);
    static firestoreNangoConnectionsSubcollectionName = "nango_connections";

    static async exists(userId: string, nangoIntegrationId: string, syncDocumentId: string): Promise<boolean> {
        const snapshot =
            await this.firestoreUsersCollection
                .doc(userId).collection(this.firestoreNangoConnectionsSubcollectionName)
                .doc(nangoIntegrationId).collection(this.firestoreCollectionName)
                .doc(syncDocumentId).get();
        return snapshot.exists;
    }

    static async get(userId: string, nangoIntegrationId: string, syncDocumentId: string): Promise<SyncGoogleMailEmail> {
        const snapshot = await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreNangoConnectionsSubcollectionName)
            .doc(nangoIntegrationId).collection(this.firestoreCollectionName)
            .doc(syncDocumentId).get();
        if (!snapshot.exists) {
            throw SemurEngineErrorBuilder.notFound("Sync information not found");
        }
        return SyncGoogleMailEmail.fromJSON(snapshot.data() as Record<string, unknown>);
    }

    static async set(userId: string, nangoIntegrationId: string, syncDocument: SyncGoogleMailEmail): Promise<void> {
        await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreNangoConnectionsSubcollectionName)
            .doc(nangoIntegrationId).collection(this.firestoreCollectionName)
            .doc(syncDocument.id).set({...syncDocument});
    }

    static async delete(userId: string, nangoIntegrationId: string, syncDocumentId: string): Promise<void> {
        await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreNangoConnectionsSubcollectionName)
            .doc(nangoIntegrationId).collection(this.firestoreCollectionName)
            .doc(syncDocumentId).delete();
    }

    static async update(userId: string, nangoIntegrationId: string, syncDocument: SyncGoogleMailEmail): Promise<void> {
        await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreNangoConnectionsSubcollectionName)
            .doc(nangoIntegrationId).collection(this.firestoreCollectionName)
            .doc(syncDocument.id).update({...syncDocument});
    }

    static async updateCustomFields(userId: string, nangoIntegrationId: string, syncDocumentId: string, data: {
        [key: string]: unknown
    }): Promise<void> {
        await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreNangoConnectionsSubcollectionName)
            .doc(nangoIntegrationId).collection(this.firestoreCollectionName)
            .doc(syncDocumentId).update(data);
    }

    static async query(
        queryKey: UserSyncGoogleMailQueryKey,
        args: UserSyncGoogleMailQueryArgs[UserSyncGoogleMailQueryKey],
    ): Promise<SyncGoogleMailEmail[]> {
        const query = await this.buildQuery(queryKey, args);
        const querySnapshot = await query.get();
        return querySnapshot.docs.map((doc) => SyncGoogleMailEmail.fromJSON(doc.data() as Record<string, unknown>));
    }

    // Build query given query key and arguments
    static async buildQuery(
        queryKey: UserSyncGoogleMailQueryKey,
        args: UserSyncGoogleMailQueryArgs[UserSyncGoogleMailQueryKey],
    ): Promise<FirebaseFirestore.Query> {
        if (!this.firestoreUsersCollection) {
            throw SemurEngineErrorBuilder.internalError("Firebase users collection is not defined");
        }
        switch (queryKey) {
            case UserSyncGoogleMailQueryKey.All: {
                const baseQuery = this.firestoreUsersCollection
                    .doc(args.userId).collection(this.firestoreNangoConnectionsSubcollectionName)
                    .doc(args.nangoIntegrationId).collection(this.firestoreCollectionName)
                    // Exclude the sync document itself
                    .where("id", "!=", this.firestoreSyncDocumentId);
                if (args.limit) {
                    return baseQuery.limit(args.limit);
                }
                return baseQuery;
            }
            case UserSyncGoogleMailQueryKey.CommonFilters: {
                const commonArgs = args as UserSyncGoogleMailQueryArgs[UserSyncGoogleMailQueryKey.CommonFilters];
                let baseQuery = this.firestoreUsersCollection
                    .doc(commonArgs.userId).collection(this.firestoreNangoConnectionsSubcollectionName)
                    .doc(commonArgs.nangoIntegrationId).collection(this.firestoreCollectionName)
                    // Exclude the sync document itself
                    .where("id", "!=", this.firestoreSyncDocumentId);
                if (commonArgs.from) {
                    baseQuery = baseQuery.where("date", ">=", commonArgs.from);
                }
                if (commonArgs.to) {
                    baseQuery = baseQuery.where("date", "<=", commonArgs.to);
                }
                if (commonArgs.threadId) {
                    baseQuery = baseQuery.where("threadId", "==", commonArgs.threadId);
                }
                baseQuery = baseQuery.orderBy("date", "desc");
                if (commonArgs.limit) {
                    return baseQuery.limit(commonArgs.limit);
                }
                return baseQuery;
            }
            default:
                throw SemurEngineErrorBuilder.internalError(`Unknown query key: ${queryKey}`);
        }
    }

    // Sync information
    static async existsSyncInfo(userId: string, nangoIntegrationId: string): Promise<boolean> {
        const snapshot =
            await this.firestoreUsersCollection
                .doc(userId).collection(this.firestoreNangoConnectionsSubcollectionName)
                .doc(nangoIntegrationId).collection(this.firestoreCollectionName)
                .doc(this.firestoreSyncDocumentId).get();
        return snapshot.exists;
    }

    static async getSyncInfo(userId: string, nangoIntegrationId: string): Promise<SyncInformation> {
        const snapshot = await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreNangoConnectionsSubcollectionName)
            .doc(nangoIntegrationId).collection(this.firestoreCollectionName)
            .doc(this.firestoreSyncDocumentId).get();
        if (!snapshot.exists) {
            throw SemurEngineErrorBuilder.notFound("Sync information not found");
        }
        return SyncInformation.fromJSON(snapshot.data() as Record<string, unknown>);
    }

    static async setSyncInfo(userId: string, nangoIntegrationId: string, sync: SyncInformation): Promise<void> {
        await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreNangoConnectionsSubcollectionName)
            .doc(nangoIntegrationId).collection(this.firestoreCollectionName)
            .doc(this.firestoreSyncDocumentId).set({...sync});
    }

    static async deleteSyncInfo(userId: string, nangoIntegrationId: string): Promise<void> {
        await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreNangoConnectionsSubcollectionName)
            .doc(nangoIntegrationId).collection(this.firestoreCollectionName)
            .doc(this.firestoreSyncDocumentId).delete();
    }

    static async updateSyncInfo(userId: string, nangoIntegrationId: string, sync: SyncInformation): Promise<void> {
        await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreNangoConnectionsSubcollectionName)
            .doc(nangoIntegrationId).collection(this.firestoreCollectionName)
            .doc(this.firestoreSyncDocumentId).update({...sync});
    }

    static async updateCustomFieldsSyncInfo(userId: string, nangoConnectionId: string, data: {
        [key: string]: unknown
    }): Promise<void> {
        await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreNangoConnectionsSubcollectionName)
            .doc(nangoConnectionId).collection(this.firestoreCollectionName)
            .doc(this.firestoreSyncDocumentId).update(data);
    }
}

File: client/semur/lib/accessors/users/syncs/user_sync_google_mail_accessor.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:semur/accessors/users/user_accessor.dart';
import 'package:semur/config/application.dart';
import 'package:semur/models.pb/syncs/sync.pb.dart';
import 'package:semur/models.pb/syncs/sync_google-mail.pbserver.dart';
import 'package:semur/transformers/syncs/sync_google_mail_email_transformer.dart';
import 'package:semur/transformers/syncs/sync_transformer.dart';

enum UserSyncGoogleMailAccessorQueryKey { ALL, LATEST_WITH_LAST_DOCUMENT }

Map<UserSyncGoogleMailAccessorQueryKey, List<String>>
userSyncGoogleMailAccessorQueryArguments = {
  UserSyncGoogleMailAccessorQueryKey.ALL: ["userId", "nangoIntegrationId"],
  UserSyncGoogleMailAccessorQueryKey.LATEST_WITH_LAST_DOCUMENT: [
    "userId",
    "nangoIntegrationId",
    "lastDocument",
  ],
};

class UserSyncGoogleMailAccessor {
  static String firebaseCollectionName = "syncs";
  static String firestoreNangoConnectionsSubcollectionName =
      "nango_connections";
  static String firestoreSyncDocumentId = "sync";

  final FirebaseFirestore firestore;
  UserSyncGoogleMailAccessor({required this.firestore});

  Query buildQuery({
    required UserSyncGoogleMailAccessorQueryKey queryKey,
    Map<String, dynamic> arguments = const {},
  }) {
    // Validate required arguments
    List<String> requiredArgs =
        userSyncGoogleMailAccessorQueryArguments[queryKey] ?? [];
    for (String arg in requiredArgs) {
      if (!arguments.containsKey(arg)) {
        throw Exception("Missing required argument '$arg' for query $queryKey");
      }
    }

    // Build query based on key
    switch (queryKey) {
      case UserSyncGoogleMailAccessorQueryKey.ALL:
        return firestore
            .collection(UserAccessor.firebaseCollectionName)
            .doc(arguments['userId'])
            .collection(firestoreNangoConnectionsSubcollectionName)
            .doc(arguments['nangoIntegrationId'])
            .collection(firebaseCollectionName)
            .where("id", isNotEqualTo: firestoreSyncDocumentId);
      case UserSyncGoogleMailAccessorQueryKey.LATEST_WITH_LAST_DOCUMENT:
        Query query = firestore
            .collection(UserAccessor.firebaseCollectionName)
            .doc(arguments['userId'])
            .collection(firestoreNangoConnectionsSubcollectionName)
            .doc(arguments['nangoIntegrationId'])
            .collection(firebaseCollectionName)
            .orderBy("date", descending: true);
        if (arguments['lastDocument'] != null &&
            arguments['lastDocument'] is DocumentSnapshot) {
          query = query.startAfterDocument(arguments['lastDocument']);
        }
        query = query.limit(50);
        return query;
    }
  }

  Future<List<SyncGoogleMailEmail>> getQuery({
    required PermissionRoles callerRole,
    required UserSyncGoogleMailAccessorQueryKey queryKey,
    Map<String, dynamic> arguments = const {},
    int? limit,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.hostUser,
      PermissionRoles.businessUser,
      PermissionRoles.user,
    ];

    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to access this data.");
    }

    try {
      Query query = buildQuery(queryKey: queryKey, arguments: arguments);
      if (limit != null) {
        query = query.limit(limit);
      }
      QuerySnapshot snapshot = await query.get();
      return snapshot.docs
          .where((doc) => doc.id != firestoreSyncDocumentId)
          .map((doc) => SyncGoogleMailEmailTransformer.fromFirebase(doc))
          .toList();
    } catch (e) {
      throw Exception("Error fetching syncs: ${e.toString()}");
    }
  }

  Future<(List<SyncGoogleMailEmail>, DocumentSnapshot?)>
  getQueryWithLastDocumentSnapshot({
    required PermissionRoles callerRole,
    required UserSyncGoogleMailAccessorQueryKey queryKey,
    Map<String, dynamic> arguments = const {},
    int? limit,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.hostUser,
      PermissionRoles.businessUser,
      PermissionRoles.user,
    ];

    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to access this data.");
    }

    try {
      Query query = buildQuery(queryKey: queryKey, arguments: arguments);
      if (limit != null) {
        query = query.limit(limit);
      }
      QuerySnapshot snapshot = await query.get();
      return (
        snapshot.docs
            .map((doc) => SyncGoogleMailEmailTransformer.fromFirebase(doc))
            .toList(),
        snapshot.docs.isNotEmpty ? snapshot.docs.last : null,
      );
    } catch (e) {
      throw Exception("Error fetching syncs: ${e.toString()}");
    }
  }

  Future<int> countQuery({
    required PermissionRoles callerRole,
    required UserSyncGoogleMailAccessorQueryKey queryKey,
    Map<String, dynamic> arguments = const {},
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.hostUser,
      PermissionRoles.businessUser,
      PermissionRoles.user,
    ];

    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to access this data.");
    }

    try {
      AggregateQuerySnapshot snapshot =
          await buildQuery(
            queryKey: queryKey,
            arguments: arguments,
          ).count().get();
      return snapshot.count ?? 0;
    } catch (e) {
      throw Exception("Error fetching syncs: ${e.toString()}");
    }
  }

  Future<SyncGoogleMailEmail> get({
    required String userId,
    required String nangoIntegrationId,
    required String syncId,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.hostUser,
      PermissionRoles.businessUser,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to access this data.");
    }
    try {
      DocumentSnapshot<Map<String, dynamic>> snapshot =
          await firestore
              .collection(UserAccessor.firebaseCollectionName)
              .doc(userId)
              .collection(firestoreNangoConnectionsSubcollectionName)
              .doc(nangoIntegrationId)
              .collection(firebaseCollectionName)
              .doc(syncId)
              .get();

      if (!snapshot.exists) {
        throw Exception("Sync email not found.");
      }
      return SyncGoogleMailEmailTransformer.fromFirebase(snapshot);
    } catch (e) {
      throw Exception("Error fetching sync email: ${e.toString()}");
    }
  }

  Future<void> set({
    required String userId,
    required String nangoIntegrationId,
    required SyncGoogleMailEmail syncEmail,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to save this data.");
    }
    try {
      await firestore
          .collection(UserAccessor.firebaseCollectionName)
          .doc(userId)
          .collection(firestoreNangoConnectionsSubcollectionName)
          .doc(nangoIntegrationId)
          .collection(firebaseCollectionName)
          .doc(syncEmail.id)
          .set(
            SyncGoogleMailEmailTransformer.toJson(syncEmail),
            SetOptions(merge: true),
          );
    } catch (e) {
      throw Exception("Error saving sync email: ${e.toString()}");
    }
  }

  Future<void> delete({
    required String userId,
    required String nangoIntegrationId,
    required String syncId,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [PermissionRoles.admin];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to delete this data.");
    }
    try {
      await firestore
          .collection(UserAccessor.firebaseCollectionName)
          .doc(userId)
          .collection(firestoreNangoConnectionsSubcollectionName)
          .doc(nangoIntegrationId)
          .collection(firebaseCollectionName)
          .doc(syncId)
          .delete();
    } catch (e) {
      throw Exception("Error deleting sync email: ${e.toString()}");
    }
  }

  Future<void> update({
    required String userId,
    required String nangoIntegrationId,
    required SyncGoogleMailEmail syncEmail,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to update this data.");
    }
    try {
      await firestore
          .collection(UserAccessor.firebaseCollectionName)
          .doc(userId)
          .collection(firestoreNangoConnectionsSubcollectionName)
          .doc(nangoIntegrationId)
          .collection(firebaseCollectionName)
          .doc(syncEmail.id)
          .update(SyncGoogleMailEmailTransformer.toJson(syncEmail));
    } catch (e) {
      throw Exception("Error updating sync email: ${e.toString()}");
    }
  }

  Future<void> updateCustomFields({
    required String userId,
    required String nangoIntegrationId,
    required String syncId,
    required Map<String, dynamic> updatedData,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to update this data.");
    }
    try {
      await firestore
          .collection(UserAccessor.firebaseCollectionName)
          .doc(userId)
          .collection(firebaseCollectionName)
          .doc(nangoIntegrationId)
          .collection(firebaseCollectionName)
          .doc(syncId)
          .update(updatedData);
    } catch (e) {
      throw Exception("Error updating sync email: ${e.toString()}");
    }
  }

  Future<bool> exists({
    required String userId,
    required String nangoIntegrationId,
    required String syncId,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.hostUser,
      PermissionRoles.businessUser,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to access this data.");
    }
    try {
      DocumentSnapshot<Map<String, dynamic>> snapshot =
          await firestore
              .collection(UserAccessor.firebaseCollectionName)
              .doc(userId)
              .collection(firestoreNangoConnectionsSubcollectionName)
              .doc(nangoIntegrationId)
              .collection(firebaseCollectionName)
              .doc(syncId)
              .get();

      return snapshot.exists;
    } catch (e) {
      throw Exception("Error checking sync email existence: ${e.toString()}");
    }
  }

  // Sync information
  Future<SyncInformation> getSyncInfo({
    required String userId,
    required String nangoIntegrationId,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.hostUser,
      PermissionRoles.businessUser,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to access this data.");
    }
    try {
      DocumentSnapshot<Map<String, dynamic>> snapshot =
          await firestore
              .collection(UserAccessor.firebaseCollectionName)
              .doc(userId)
              .collection(firestoreNangoConnectionsSubcollectionName)
              .doc(nangoIntegrationId)
              .collection(firebaseCollectionName)
              .doc("sync")
              .get();

      if (!snapshot.exists) {
        throw Exception("Sync info not found.");
      }
      return SyncTransformer.fromFirebase(snapshot);
    } catch (e) {
      throw Exception("Error fetching sync info: ${e.toString()}");
    }
  }
}
